cpp/src/io/parquet/reader_impl.cu

```diff
@@ -1213,7 +1209,9 @@ rmm::device_buffer reader::impl::decompress_page_data(
     start_pos += codec.num_pages;
   }
 
-  decompress_check(comp_stats, stream);
+  decompress_check(comp_stats, any_block_failure.device_ptr(), stream);
+  any_block_failure.device_to_host(stream, true);
+  CUDF_EXPECTS(not any_block_failure[0], "Error during decompression");
```

Comment:
I don't understand how this is avoiding a synchronization. It's just moving the sync from being implicit in the algorithm to explicit.

Reply:
My understanding of the issue is that `any_of` does a device level synchronize. Does `count_if` behave differently?

Reply:
It shouldn't be any different than any other Thrust algorithm when you pass it a stream.

Reply:
> I don't understand how this is avoiding a synchronization

This uses a copy to pinned memory so synchronizes the stream instead of the device. @cheinger found that the thrust algorithms he tried all involved a device synchronize to copy the result to be returned.

Reply:
Isn't cub device reduce guaranteed to not sync because it doesn't have to return. It can just let the value stay in device memory, which we can copy over later. I don't know how much preferred this is to the hand rolled kernel.

Reply:
> This uses a copy to pinned memory so synchronizes the stream instead of the device.

Pinned or pageable memory should not impact the behavior of a stream synchronize.

Reply:
Reimplemented with thrust and cub. Bit on the verbose side, but it should be the best option: no custom kernel with UB + result in pinned memory.
Open to suggestions on ways to avoid `thrust::transform`.

Reply:
I'm really sure how, but the original change to move away from `thrust::equal` improved performance in the Spark use case.
CC @cheinger to help explain.

Reply:
To clarify, the issue isn't with the synchronization itself, but rather the location of the synchronization.

When I made the original change to replace the `thrust::equal` with a kernel, it was so that I could delay the host side evaluation of nvCOMP decompression statuses until later so that the host synchronization didn't block kernels from being overlapped.

This is what the TPC-DS Q9 profile looked like when synchronizes were in the wrong place:
<img width="922" alt="Screen Shot 2022-07-21 at 3 33 38 PM" src="https://user-images.githubusercontent.com/12998801/180326647-29ee0e6a-8569-4fc6-9b78-19e20af0f833.png">

Vs here when they were removed/delayed:
<img width="922" alt="Screen Shot 2022-07-21 at 3 33 48 PM" src="https://user-images.githubusercontent.com/12998801/180326774-f114d385-d5f7-49dc-9393-2a7f1ae5a17e.png">

@jrhemstad is correct in saying that cudaHostAlloc and cudaFree will cause device synchronizes, and this was also affecting the kernel overlap. When I last worked on this, these host allocations were affecting the kernel overlap as well so I moved them around in such a way I could yield better overlap which resulted better query performance (I think it was somewhere between 10-25%).

Reply:
To clarify, the issue isn't with the synchronization itself, but rather the **location** of the synchronization.

When I made the original change to replace the `thrust::equal` with a kernel, it was so that I could delay the host side evaluation of nvCOMP decompression statuses until later so that the host synchronization didn't block kernels from being overlapped.

This is what the TPC-DS Q9 profile looked like when synchronizes were in the wrong place:
<img width="922" alt="Screen Shot 2022-07-21 at 3 33 38 PM" src="https://user-images.githubusercontent.com/12998801/180326647-29ee0e6a-8569-4fc6-9b78-19e20af0f833.png">

Vs here when they were removed/delayed:
<img width="922" alt="Screen Shot 2022-07-21 at 3 33 48 PM" src="https://user-images.githubusercontent.com/12998801/180326774-f114d385-d5f7-49dc-9393-2a7f1ae5a17e.png">

@jrhemstad is correct in saying that cudaHostAlloc and cudaFree will cause device synchronizes. When I last worked on this, specifically looking at TPC-DS Q9, these host allocations were affecting the kernel overlap as well, so I moved them around in such a way that I could yield better overlap which resulted in better query performance. I never submitted a MR for this allocation optimization because they were going to be replaced with an allocator.

I think the key take away is to be careful where implicit/explicit host synchronizes are placed, because they can affect how your CUDA kernels are overlapped.